### PR TITLE
side-effects refactor

### PR DESCRIPTION
**Pull request by jhamman**

### CODE REVIEW

**Comment by norlandrhagen** [comment]
### General 

- [ ] update `rechunk`
- [ ] update `regrid`

### bcsd

- [ ] rename `methods/bcsd.py` to `methods/bcsd/core.py` #118 

#### bcsd flow

- [ ] `coarse_obs`

File: cmip6_downscaling/methods/common/containers.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    method: str
+    obs: str
+    model: str
+    scenario: str
+    variable: str
+    latmin: str
+    latmax: str
+    lonmin: str
+    lonmax: str
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/containers.py:60]: should these be floats?

Comment (norlandrhagen) [reply cmip6_downscaling/methods/common/containers.py]: https://github.com/carbonplan/cmip6-downscaling/pull/140

File: cmip6_downscaling/__init__.py [hunk 2]
```diff
@@ -1,8 +1,15 @@
 from dask.utils import SerializableLock
 from donfig import Config
+from pkg_resources import DistributionNotFound, get_distribution
 
 from cmip6_downscaling.config import _defaults
```
Comment (andersy005) [inline cmip6_downscaling/__init__.py:5]:
```suggestion
from .config import _defaults
```

File: cmip6_downscaling/methods/bcsd/flow.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    annual_summary,
+    get_experiment,
+    get_obs,
+    make_run_parameters,
+    monthly_summary,
+    pyramid,
+    rechunk,
+    regrid,
+    run_analyses,
+)
```
Comment (andersy005) [inline cmip6_downscaling/methods/bcsd/flow.py:21]:
```suggestion
from . import runtimes
from .methods.bcsd.tasks import (
    fit_and_predict,
    postprocess_bcsd,
    spatial_anomalies,
)
from .methods.common.tasks import (
    annual_summary,
    get_experiment,
    get_obs,
    make_run_parameters,
    monthly_summary,
    pyramid,
    rechunk,
    regrid,
    run_analyses,
)
```

File: cmip6_downscaling/methods/bcsd/tasks.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import xarray as xr
+from prefect import task
+from skdownscale.pointwise_models import PointWiseDownscaler
+from skdownscale.pointwise_models.bcsd import BcsdPrecipitation, BcsdTemperature
+from upath import UPath
+
+from cmip6_downscaling import config
+from cmip6_downscaling.constants import ABSOLUTE_VARS, RELATIVE_VARS
+from cmip6_downscaling.methods.common.containers import RunParameters
+from cmip6_downscaling.methods.common.utils import zmetadata_exists
```
Comment (andersy005) [inline cmip6_downscaling/methods/bcsd/tasks.py:13]:
```suggestion
from . import config
from .constants import ABSOLUTE_VARS, RELATIVE_VARS
from .methods.common.containers import RunParameters
from .methods.common.utils import zmetadata_exists
```

File: cmip6_downscaling/methods/common/utils.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+from .containers import BBox
+
+
+def zmetadata_exists(path: UPath):
+    '''temporary workaround until path.exists() works'''
+
+    if isinstance(path, pathlib.PosixPath):
+        return (path / '.zmetadata').exists()
+    else:
+        return (UPath(path) / '.zmetadata').exists()
```
Comment (jhamman) [inline cmip6_downscaling/methods/common/utils.py:20]:
```suggestion
    if isinstance(path, pathlib.PosixPath):
        return (path / '.zmetadata').exists()
    elif isinstance(path, UPath):
        return path.fs.exists(str(path / '.zmetadata'))
    else:
        return (UPath(path) / '.zmetadata').exists()
```

**Pull request merged**